mds/QuiesceDb: Manager, Agent, a-sock i-face, mgr/volumes CLI (#54485)
ceph/ceph: doc/cephfs/fs-volumes.rst

   ceph fs subvolume resize <vol_name> <subvol_name> <new_size> [--group_name <subvol_group_name>] [--no_shrink]

The command resizes the subvolume quota using the size specified by ``new_size``.

> The ``--no_shrink`` flag prevents the subvolume from shrinking below the current  used size of the subvolume.

Reviewer: nit:  double space

group.  This will cause every subvolume within the group to be automatically
pinned to one of the available ranks on the file system.

Subvolume quiesce
-----------------

> It may be needed to pause (write) IO to a set of subvolumes of a given file system.

Reviewer 2: ```suggestion
It may be needed to pause (write) IO to a set of subvolumes of a given volume (file system).
```

A good example of such case is a consistent snapshot spanning multiple subvolumes.
Such a task arises often in an environment such as k8s, where a single deployed application
can work with many mounted subvolumes across several hosts. When a snapshot of such a system is needed,
the application may not find the result consistent unless the snapshots were taken
under an active write pause.

The volumes plugin provides a tool to initiate and await such a pause across a set of subvolumes:

.. prompt:: bash $ auto

  $ ceph fs subvolume quiesce --set-id myset1 <vol_name> <[group_name/]sub_name...> --await
  # perform actions while the IO pause is active, like taking snapshots
  $ ceph fs subvolume quiesce --set-id myset1 --release --await
  # if successful, all members of the set were confirmed as still in pause and released from such

The ``subvolume quiesce`` functionality is itself based on top of a lower level QuiesceDb service maintained by the MDS
daemons. Volumes plugin merely maps the subvolume names to their corresponding paths on the given file system

> and then issues the corresponding quiesce db command. You can learn more about the feature in the developer guides.

Reviewer 2: ```suggestion
and then issues the corresponding quiesce command to the MDS. You can learn more about the feature in the developer guides.
```

Operations
^^^^^^^^^^

The IO pause (referred to as `quiesce`) is requested for a group of one or more subvolumes (i.e. paths in a filesystem).
The group is referred to as "quiesce set", and every quiesce set must have a unique string id to interact with.
A quiesce set can be manipulated in the following ways:

* **include** one or more subvolumes - quiesce set members
* **exclude** one or more members
* **cancel** the set, asynchronously aborting the pause on all its current members
* **release** the set, requesting the end of the pause from all members and expecting an ack from all clients
* **query** the current state of a set by id or all active sets or all known sets
* **cancel all** active sets in case an immediate resume of IO is required.
* **await** the set to change states, applicable as a modifier to other requests

> All of the operations listed above are carried out syncronously by the designated quiesce leader,

Reviewer 2: ```suggestion
All of the operations listed above are carried out asynchronously by the designated quiesce leader,
```

?

Author: Probably not the best wording. I did intend to use `synchronously`, but I understand your logic here.

So, what I'm trying to say is that when `--await` or `--await-for` is not used, then the operations perform their designated changes and return with the latest version of the set they were targeted at. So, I can't call it `asynchronous` since they do pend on the quiesce db manager to apply them in the next batch. However, it shouldn't ever take long - the delays are just the transport and quiesce db manager control loop latency.

Reviewer 2: Okay. In any case the spelling needs fixed :)

Reviewer: That's very Pittsburghese of you ;)

Author: I do see how this was confusing for you though. I will appreciate suggestions to make it clearer.
I will probably just have to add more words, along the lines of my previous comment

                      }
                  }
              }
          }
      }
  }

The ``--include`` bit is optional, as if no operation is given while members are provided, 
then "include" is assumed.

As we have seen, the timeout argument specifies how much time we are ready to give the system
to reach the `QUIESCED` state on the set. However, since new members can be added to an
active set at any time, it wouldn't be fair to measure the timeout from the set creation time.
Hence, the timeout is tracked per member: every member has `timeout` seconds to quiesce,
and if any one takes longer than that, the whole set is marked as `TIMEDOUT` and the pause is released.

Once the set is in the `QUIESCED` state, it will begin its expiration timer. This timer is tracked
per set as a whole, not per members. Once the `expiration` seconds elapse, the set will transition
into an `EXPIRED` state, unless it was actively released or canceled by a dedicated operation.

It's possible to add new members to a `QUIESCED` set. In this case, it will transition back to `QUIESCING`,
and the new member(s) will have their own timeout to quiesce. If they succeed, then the set will

> again be `QUIESCED` and the expiration timer will restart. The expiration timer doesn't apply
> when a set is `QUIESCING`

Reviewer 2: ```suggestion
again be `QUIESCED` and the expiration timer will restart. **The expiration timer doesn't apply
when a set is `QUIESCING`**.
```

I would emphasize this bit loudly.

Awaiting
^^^^^^^^

Note that the commands above are all non-blocking. If we want to wait for the quiesce set

> to reach the `QUIESCED` state, we should await it at some point. ``--await`` can be givent

Reviewer 2: ```suggestion
to reach the `QUIESCED` state, we should await it at some point. ``--await`` can be given
```

along with other arguments to let the system know our intention.

Technically, there are two types of await: `quiesce await` and `release await`. The former
is the default, and the latter can only be achieved with ``--release`` present in the argument list.
To avoid confision, it is not permitted to issue a `quiesce await` when the set is already `RELEASING`
or `RELEASED`. Trying to ``--release`` a set that is not `QUIESCED` is an ``EPERM`` error as well, regardless
of whether await is requested alongside. However, it's not an error to `release await` 
an already released set, or to `quiesce await` a `QUIESCED` one.

> When awaiting, one may also specify a maximum duration that they'd like this await request to block for,

Reviewer 2: ```suggestion
When awaiting, one may also specify a maximum duration that they would like this await request to block for,
```

suggest avoiding contractions in technical docs

While it doesn't make sense to specify both arguments, it is not considered an error. If
both ``--await`` and ``--await-for`` are present, then the former is ignored, and the time limit
from ``--await-for`` is honored.

.. prompt:: bash $ auto

  $ time ceph fs subvolume quiesce a sub1 --timeout=10 --await-for=2
  {
      "epoch": 6,
      "db_version": 3,
      "sets": {
          "c3c1d8de": {
              "db_version": 3,
              "age_ref": 0.0,
              "state": {
                  "name": "QUIESCING",
                  "age": 2.0
              },
              "timeout": 10.0,
              "expiration": 0.0,
              "members": {
                  "file:/volumes/_nogroup/sub1/b1fcce76-3418-42dd-aa76-f9076d047dd3": {
                      "excluded": false,
                      "state": {
                          "name": "QUIESCING",
                          "age": 2.0
                      }
                  }
              }
          }
      }
  }
  Error EINPROGRESS: 
  ceph fs subvolume quiesce a sub1 --timeout=10 --await-for=2  0.41s user 0.04s system 17% cpu 2.563 total

> (there is a ~0.5 sec overhead that the ceph client adds, at least in a local debug setup)

Reviewer 2: ```suggestion
..note:: There is a ~0.5 sec overhead that the ceph client adds, at least in a local debug setup.
```

the system to spend up to 30 seconds reaching the quiesced state across all members
and stay quiesced for up to 10 seconds before the quiesce expires and the IO
is resumed. We also specify ``--await`` to only proceed once the quiesce is reached.

We then proceed with a set of command pairs that take the next snapshot and call ``--await`` on our set
to extend the expiration timeout for 10 more seconds `(3,4)`. This approach gives us up to 10 seconds
for every snapshot, but also allows taking as many snapshots as we need without losing the IO pause,
and with it - consistency. If we wanted, we could update the `expiration` every time we called for await.

If any of the snapshots gets stuck and takes longer than 10 seconds to complete, then the next call
to ``--await`` will return an error since the set will be `EXPIRED` which is not an awaitable state.
This limits the impact on the applications in the bad case scenarios.

We could have set the `expiration` timeout to 30 at the beginning `(2)`, but that would mean that
a single stuck snapshot would keep the applications pending for all this time.

If Version
^^^^^^^^^^

> Sometimes, it's not enough to just observe the successful quesce or release. The reason could be

Reviewer 2: ```suggestion
Sometimes, it's not enough to just observe the successful quiesce or release. The reason could be
```

If the set is already at an ``(A)`` state, the operation completes immediately with a success. 

Most of the operations require a set-id. The exceptions are:

* creation of a new set without specifying a set id,
* query of active or all known sets, and
* the cancel all

Creating a new set is achieved by including member(s) via the `include` or `reset` commands.
It's possible to specify a set id, and if it's a new id then the set will be created
with the specified member(s) in the `QUIESCING` state. When no set id is specified while including
or resetting members, then a new set with a unique set id is created. The set id will be known
to the caller by inspecting the output

.. prompt:: bash $ auto

  $ ceph fs subvolume quiesce a sub1 --set-id=unique-id
  {
      "epoch": 3,
      "db_version": 1,
      "sets": {
          "unique-id": {
              "db_version": 1,
              "age_ref": 0.0,
              "state": {
                  "name": "TIMEDOUT",
                  "age": 0.0
              },
              "timeout": 0.0,
              "expiration": 0.0,
              "members": {

>                   "file:/volumes/_nogroup/sub1/b1fcce76-3418-42dd-aa76-f9076d047dd3": {

Reviewer 3: What does the `file:` prefix indicate? Can it ever be anything else?

Author: It can also be `inode:nnnnnnn`, but as of the current implementation we only work with `file:` roots.
The quiesce db will validate and sanitize the roots that are sent at the initial phase of request processing. If a root doesn't have a scheme then it's assumed to be an absolute path and is therefore converted to the `file:` format.

Reviewer 3: A thorough explanation! thanks!

group.  This will cause every subvolume within the group to be automatically
pinned to one of the available ranks on the file system.

Subvolume quiesce
-----------------

It may be needed to pause IO to a set of subvolumes of a given volume (file system).
A good example of such case is a consistent snapshot spanning multiple subvolumes.

> Such a task arises often in an environment such as k8s, where a single deployed application

Reviewer: suggest s/k8s/Kubernetes/

can work with many mounted subvolumes across several hosts. When a snapshot of such a system is needed,
the application may not find the result consistent unless the snapshots were taken

> under an active write pause.

Reviewer: s/under/during/ ?

> The volumes plugin provides a tool to initiate and await such a pause across a set of subvolumes:

Reviewer: backticks around ``volumes``?
